First pass at supporting structured logging for a few sources.

bmoyles0117 opened the pull request:
Exploring options for structured logging. Any feedback for the approach taken so far would be greatly appreciated as we continue to support more structured logs by default.

bmoyles0117 committed:
First pass at supporting structured logging for a few sources.

qingling128 commented on configs/config.d/apache.conf:
```diff
@@ -9,7 +9,8 @@
 <source>
   @type tail
-  format none
+  # Sample line: [Tue Oct 24 17:20:22.052283 2017] [core:notice] [pid 5997] AH00094: Command line: '/usr/sbin/apache2'
```
NIT: Can we put "Sample line: " and the actual line in two different lines? Similar for other comments below.

bmoyles0117 replied:
Done.

qingling128 commented on configs/config.d/gitlab.conf:
```diff
@@ -1,3 +1,5 @@
+# None of these files existed when installing gitlab...
+
```
Seems like the log files are in these location only when we install from source. Otherwise they are at: `/var/log/gitlab/gitlab-rails`. See reference at https://docs.gitlab.com/ee/administration/logs.html.

We could install gitlab via source then check the log format from there instead.

bmoyles0117 replied:
Implemented! Thanks for the pointer 👍

qingling128 commented on configs/config.d/jenkins.conf:
```diff
@@ -1,8 +1,11 @@
   format multiline
   format_firstline /^\w+\s\d+,\s\d+/
-  format1 /(?<message>.*)/
+  format1 /(?<verbose_time>\w+ [0-9]+, [0-9]+ [0-9]+:[0-9]+:[0-9]+ \w+) (?<class>[^\s]+) (?<method>\w+)\n/
```
- Why do we call this `verbose_time` instead of `time`? The `fluent-plugin-google-cloud` knows to extract timestamp from `time`, but not from `verbose_time`. If possible, we should stick with format that will be understood by `fluent-plugin-google-cloud`.

- Is the `> ` in the sample part of the log? If so, that should be reflected in the regex as well, right?

- The regex for `format_firstline` seems a bit loose:
`format_firstline /^\w+\s\d+,\s\d+/`. I wonder if it might match other patterns in regular logs by accident. It might be better to use a more strict pattern, similar to `verbose_time`. Something like:
```
/^\w+ \d{1,2}, \d{1,4} \d{1,2}:\d{1,2}:\d{1,2} \w+/
```
- Replace [0-9] with \d?

qingling128 commented on configs/config.d/jenkins.conf:
```diff
@@ -1,8 +1,11 @@
   format_firstline /^\w+\s\d+,\s\d+/
-  format1 /(?<message>.*)/
+  format1 /(?<verbose_time>\w+ [0-9]+, [0-9]+ [0-9]+:[0-9]+:[0-9]+ \w+) (?<class>[^\s]+) (?<method>\w+)\n/
+  format2 /(?<level>[^:]+):\s+(?<message>.*)\n/
```
- Is the `>` in the sample part of the log? If so, that should be reflected in the regex as well, right?

- It might seem more natural to match `$` at the end instead of `\n`.

bmoyles0117 replied:
The `>` is not part of the log line, I was trying to visualize common indentation, I've updated it to be below "Sample format". It also seems like \n is what they use in the examples specifically when using multi line formats.

qingling128 replied:
I see. Thanks for the clarification. Let's stick with `\n` then.

qingling128 commented on configs/config.d/cassandra.conf:
```diff
@@ -1,6 +1,8 @@
   @type tail
-  format none
+  # Sample line: INFO  [main] 2017-10-23 22:55:47,737 TokenMetadata.java:479 - Updating topology for localhost/127.0.0.1
+  format /^(?<level>[^\s]+)\s+\[(?<package>[^\]]+)] (?<time>[0-9]{1,4}-[0-9]{1,2}-[0-9]{1,2} [0-9]{1,2}:[0-9]{1,2}:[0-9]{1,2},[0-9]{1,3}) (?<java_file>[^\:]+):(?<line_number>\d+) - (?<message>.*)$/
```
Replace `[0-9]` with `\d`?

bmoyles0117 replied:
Done.

qingling128 commented on configs/config.d/mongodb.conf:
```diff
@@ -1,6 +1,9 @@
-  format none
+
+  # Sample line: 2017-10-27T20:39:28.924+0000 I JOURNAL  [initandlisten] preallocateIsFaster=true 2.82
+  format /^(?<time>[0-9]{4}-[0-9]{2}-[0-9]{2}T[0-9]{1,2}:[0-9]{2}:[0-9]{2}\.[0-9]{3}[\+-][0-9]{4})\s+(?<severity>\w+)\s+(?<component>[\w\s]+)\s+\[(?<context>[\w\s]+)\] (?<message>.*)$/
```
- Replace [0-9] with \d?

- `(?<component>[\w\s]+)`? Do you mean `(?<component>\w+)`? Similarly, for `(?<context>[\w\s]+)`, are we really expecting white spaces?

bmoyles0117 replied:
Done for digits, we are indeed expecting white spaces in this context. I'll try to use another example that demonstrates the significance of the whitespace.

qingling128 commented on configs/config.d/redis.conf:
```diff
@@ -1,5 +1,6 @@
 <source>
   @type tail
+  # The format across lines is not consistent, can not deduce structure.
```
🤦‍♀️  What's that like? I saw a blog post: http://build47.com/redis-log-format-levels/. Not sure how useful it would be though.

bmoyles0117 replied:
Some log lines include the log level, and others don't. I'll find the example and paste it here.

bmoyles0117 replied:
This is what the log file contains. WDYT?

```23439:C 30 Oct 15:33:59.003 # oO0OoO0OoO0Oo Redis is starting oO0OoO0OoO0Oo
23439:C 30 Oct 15:33:59.004 # Redis version=4.0.2, bits=64, commit=00000000, modified=0, pid=23439, just started
23439:C 30 Oct 15:33:59.004 # Configuration loaded
23440:M 30 Oct 15:33:59.006 * Increased maximum number of open files to 10032 (it was originally set to 1024).
                _._                                                  
           _.-``__ ''-._                                             
      _.-``    `.  `_.  ''-._           Redis 4.0.2 (00000000/0) 64 bit
  .-`` .-```.  ```\/    _.,_ ''-._                                   
 (    '      ,       .-`  | `,    )     Running in standalone mode
 |`-._`-...-` __...-.``-._|'` _.-'|     Port: 6379
 |    `-._   `._    /     _.-'    |     PID: 23440
  `-._    `-._  `-./  _.-'    _.-'                                   
 |`-._`-._    `-.__.-'    _.-'_.-'|                                  
 |    `-._`-._        _.-'_.-'    |           http://redis.io        
  `-._    `-._`-.__.-'_.-'    _.-'                                   
 |`-._`-._    `-.__.-'    _.-'_.-'|                                  
 |    `-._`-._        _.-'_.-'    |                                  
  `-._    `-._`-.__.-'_.-'    _.-'                                   
      `-._    `-.__.-'    _.-'                                       
          `-._        _.-'                                           
              `-.__.-'                                               

23440:M 30 Oct 15:33:59.007 # WARNING: The TCP backlog setting of 511 cannot be enforced because /proc/sys/net/core/somaxconn is set to the lower value of 128.
23440:M 30 Oct 15:33:59.007 # Server initialized
23440:M 30 Oct 15:33:59.007 # WARNING overcommit_memory is set to 0! Background save may fail under low memory condition. To fix this issue add 'vm.overcommit_memory = 1' to /etc/sysctl.conf and then reboot or run the command 'sysctl vm.overcommit_memory=1' for this to take effect.
23440:M 30 Oct 15:33:59.007 # WARNING you have Transparent Huge Pages (THP) support enabled in your kernel. This will create latency and memory usage issues with Redis. To fix this issue run the command 'echo never > /sys/kernel/mm/transparent_hugepage/enabled' as root, and add it to your /etc/rc.local in order to retain the setting after a reboot. Redis must be restarted after THP is disabled.
23440:M 30 Oct 15:33:59.007 * Ready to accept connections
23440:M 30 Oct 15:49:00.022 * 1 changes in 900 seconds. Saving...
23440:M 30 Oct 15:49:00.023 * Background saving started by pid 12431
12431:C 30 Oct 15:49:00.027 * DB saved on disk
12431:C 30 Oct 15:49:00.027 * RDB: 6 MB of memory used by copy-on-write
23440:M 30 Oct 15:49:00.123 * Background saving terminated with success```

qingling128 commented on configs/config.d/redmine.conf:
```diff
@@ -1,6 +1,7 @@
 <source>
   @type tail
   format none
+  # /var/log/redmine/default/production.log was empty after running 5 minutes.
```
Interesting.

Looking at a few `production.log` samples in their forum, there are some patterns. But they are mixed together.

http://www.redmine.org/boards/2/topics/51723
http://www.redmine.org/boards/2/topics/51978

```
Started GET "/projects/messagerie-test/issues" for 192.168.10.20 at 2017-01-03 10:34:06 +0100
Processing by IssuesController#index as HTML
Parameters: {"project_id"=>"messagerie-bluemind"}
Current user: georges.robert (id=129)
Completed 500 Internal Server Error in 70ms (ActiveRecord: 28.6ms)

NoMethodError (undefined method `each' for nil:NilClass):
app/models/issue_query.rb:357:in `issues'
app/controllers/issues_controller.rb:70:in `index'
lib/redmine/sudo_mode.rb:63:in `sudo_mode'
plugins/redmine_dmsf/lib/redmine_dmsf/webdav/no_parse.rb:33:in `call'
```

```
Started GET "/redmine/users/531" for 172.18.140.18 at 2017-02-06 13:37:42 +0300
Processing by UsersController#show as HTML
Parameters: {"id"=>"531"}
Current user: seckin.aktas (id=531)
Rendered users/show.html.erb within layouts/base (101.6ms)
Completed 200 OK in 209ms (Views: 115.8ms | ActiveRecord: 29.5ms)
Started GET "/redmine/users/531/edit" for 172.18.140.18 at 2017-02-06 13:37:55 +0300
Processing by UsersController#edit as HTML
Parameters: {"id"=>"531"}
Current user: seckin.aktas (id=531)
Rendered plugins/redmine_digest/app/views/account_settings/_settings_2x.html.erb (1.8ms)
Rendered plugins/redmine_landing_page/app/views/users/_landing_page.html.erb (0.4ms)
Rendered users/_mail_notifications.html.erb (48.8ms)
Rendered users/_preferences.html.erb (25.1ms)
Rendered users/_form.html.erb (86.3ms)
Rendered users/_general.html.erb (87.5ms)
Rendered users/_groups.html.erb (7.9ms)
Rendered principal_memberships/_index.html.erb (298.7ms)
Rendered users/_memberships.html.erb (299.1ms)
Rendered common/_tabs.html.erb (396.5ms)
Rendered users/edit.html.erb within layouts/admin (399.9ms)
Rendered admin/_menu.html.erb (4.2ms)
Rendered layouts/base.html.erb (29.3ms)
Completed 200 OK in 443ms (Views: 431.9ms | ActiveRecord: 6.1ms)

Started PATCH "/redmine/users/531" for 172.18.140.18 at 2017-02-06 13:38:05 +0300
Processing by UsersController#update as HTML
Parameters: {"utf8"=>"✓", "authenticity_token"=>"KaL/dY5t2nARw3achYyxod7M6nlijTPovob0jv7YcspmWiBio9kQW29IpZGLxWt46USpHafXpweRyfOll4E6Xg==", "user"=>{"login"=>"seckin.aktas", "firstname"=>"Seckin", "lastname"=>"Aktas", "mail"=>"xxxx", "language"=>"tr", "digest_account_attributes"=>{"active"=>"1", "id"=>"61"}, "landing_page"=>"", "auth_source_id"=>"1", "password"=>"[FILTERED]", "password_confirmation"=>"[FILTERED]", "generate_password"=>"[FILTERED]", "must_change_passwd"=>"0", "mail_notification"=>"only_my_events", "notified_project_ids"=>[""]}, "pref"=>{"no_self_notified"=>"0", "hide_mail"=>"1", "time_zone"=>"Istanbul", "comments_sorting"=>"asc", "warn_on_leaving_unsaved"=>"1"}, "commit"=>"Kaydet", "id"=>"531"}
Current user: seckin.aktas (id=531)
Completed 500 Internal Server Error in 11ms (ActiveRecord: 3.0ms)

ActiveModel::ForbiddenAttributesError (ActiveModel::ForbiddenAttributesError):
lib/redmine/safe_attributes.rb:84:in `safe_attributes='
app/controllers/users_controller.rb:135:in `update'
lib/redmine/sudo_mode.rb:63:in `sudo_mode'
```

I'm fine if we leave it to the users to decide exactly how to handle this.

bmoyles0117 committed:
Address feedback for [0-9] -> \d regex updates and sample lines.

bmoyles0117 committed:
Updating the gitlab structured logging thanks to @qingling128.

Reordered the sources to match the ordering from
https://docs.gitlab.com/ee/administration/logs.html

bmoyles0117 committed:
Supporting SugarCRM.

bmoyles0117 committed:
Documenting why we don't support structured logging for redmine.

bmoyles0117 committed:
Apache needs a time_format set to satisfy its default format.

bmoyles0117 committed:
Fixing log level field name in nginx structured logs.

bmoyles0117 committed:
Adding structured logging for some of the Chef log files.

qingling128 commented on configs/config.d/gitlab.conf:
```diff
@@ -12,34 +42,49 @@
-  path /home/git/gitlab/log/sidekiq.log,
+  # Sample format
+  # 2014-06-10T07:55:20Z 2037 TID-tm504 ERROR: /opt/bitnami/apps/discourse/htdocs/vendor/bundle/ruby/1.9.1/gems/redis-3.0.7/lib/redis/client.rb:228:in `read'
+  format /^[^\s]+\s+(?<time>[^\s]+)\s+(?<pid>\d+)\s+(?<tid>[^\s]+)\s+(?<level>[^:]+):\s+(?<message>.*)$/
```
- What is the first part `[^\s]+\s+` trying to match? The sample seems to suggest that `time` is at the beginning of the line.

qingling128 commented on configs/config.d/gitlab.conf:
```diff
@@ -12,34 +42,49 @@
+<source>
+  @type tail
+  # Sample format
+  # I, [2015-02-13T06:17:00.671315 #9291]  INFO -- : Adding project root/example.git at </var/opt/gitlab/git-data/repositories/root/dcdcdcdcd.git>.
```
Just cursious: what is `I, ` at the beginning of the line?

qingling128 commented on configs/config.d/gitlab.conf:
```diff
@@ -48,28 +93,19 @@
 <source>
   @type tail
   format none
```
Can we document the reason we don't provide a default `format` for this log?

qingling128 commented on configs/config.d/gitlab.conf:
```diff
@@ -48,28 +93,19 @@
 
 <source>
   @type tail
   format none
```
Can we document the reason we don't provide a default `format` for this log?

qingling128 commented on configs/config.d/chef.conf:
```diff
@@ -74,7 +76,10 @@
 
 <source>
```
Add a sample format here?

bmoyles0117 committed:
Adding auth log to default structured logging. (#17)

bmoyles0117 committed:
Change package names for deb and rpm.

bmoyles0117 committed:
Update the package maintainer.

bmoyles0117 committed:
Update the version to reflect new sequencing.

jkohen commented:
@bmoyles0117 what's the status of this feature?

bmoyles0117 commented:
I believe this has been superseded by a different approach for structured logging, @qingling128 can you confirm?

qingling128 commented:
Right. That is obsolete. We decided to use community-owned Fluentd parsers (https://docs.fluentd.org/parser) instead of maintaining the regex per application. See more details in https://cloud.google.com/logging/docs/structured-logging.
